File: scala-3/euler/src/main/scala/com/rbleggi/euler/P30.scala
package com.rbleggi.euler

@main def mainP30(): Unit = {
  val power = 5
  val upper = 6 * math.pow(9, power).toInt

  val result = (2 to upper).filter { n =>
    n == n.toString.map(_.asDigit).map(d => math.pow(d, power).toInt).sum
  }.sum

  println(s"Sum of all numbers that can be written as the sum of fifth powers of their digits: $result")
}

File: kotlin/euler/src/main/kotlin/com/rbleggi/euler/P30.kt
package com.rbleggi.euler

import kotlin.math.pow

fun mainP30() {
    val power = 5
    val upper = (6 * 9.0.pow(power)).toInt()

    val result = (2..upper).filter { n ->
        n == n.toString().sumOf { it.digitToInt().toDouble().pow(power).toInt() }
    }.sum()

    println("Sum of all numbers that can be written as the sum of fifth powers of their digits: $result")
}
